### PR TITLE
Require presence array

### PR DESCRIPTION
**Pull request by Graziel**
as discussed in #8581

### CODE REVIEW

File: src/Validation/Validator.php [hunk 1]
```diff
@@ -422,9 +422,16 @@ public function remove($field, $rule = null)
...
-     * Sets whether a field is required to be present in data array.
+     * Sets whether a field is required to be present in data array. You can also pass array.
+     * Using an array will let you provide the following keys:
      *
-     * @param string $field the name of the field
+     * - `mode` individual mode for field
+     * - `message` individual error message for field
+     *
+     * You can also set mode and message for all passed fields, the individual settings
+     * takes precedence over group setting.
```
Comment (dereuromark) [inline src/Validation/Validator.php]: "settings take ... settings" maybe

File: src/Validation/Validator.php [hunk 2]
```diff
@@ -433,10 +440,30 @@ public function remove($field, $rule = null)
...
+            $field = [
+                $field => [
+                    'mode' => $mode,
+                    'message' => $message
+                ]
+            ];
         }
+
+        foreach ($field as $fieldName => $setting) {
+            if (is_string($setting)) {
```
Comment (markstory) [inline src/Validation/Validator.php]: Wouldn't `$fieldName` also be an integer in this case?

File: tests/TestCase/Validation/ValidatorTest.php [hunk 3]
```diff
@@ -192,6 +192,32 @@ public function testRequirePresence()
...
 
     /**
+     * Tests the requirePresence method
+     *
+     * @return void
+     */
+    public function testRequirePresenceAsArray()
+    {
+        $validator = new Validator;
+        $validator->requirePresence(['title', 'created']);
```
Comment (markstory) [inline tests/TestCase/Validation/ValidatorTest.php]: What should happen with 'bad' input like:

```php
requirePresence(['title' => 'derp', 'created' => false])
```

File: src/Validation/Validator.php (hunk 1 above)
Comment (Graziel) [inline src/Validation/Validator.php]: i wanted 
'the individual setting takes precedence over group settings.' 
does this make more sense / is grammatically correct?

File: src/Validation/Validator.php (hunk 2 above)
Comment (Graziel) [inline src/Validation/Validator.php]: so 
`if (is_string($setting) && is_int($fieldName)) {`
isnt that a bit redundant check?

File: tests/TestCase/Validation/ValidatorTest.php (hunk 3 above)
Comment (Graziel) [inline tests/TestCase/Validation/ValidatorTest.php]:
```
if (!is_array($setting)) {
       throw new InvalidArgumentException(sprintf('Invalid field "%s" setting, must be an array.', $fieldName));
}
```
should do the trick?

File: src/Validation/Validator.php (hunk 1 above)
Comment (dereuromark) [inline src/Validation/Validator.php]: Yes I think so.

File: tests/TestCase/Validation/ValidatorTest.php (hunk 3 above)
Comment (markstory) [inline tests/TestCase/Validation/ValidatorTest.php]: That works for me.

File: src/Validation/Validator.php (hunk 2 above)
Comment (markstory) [inline src/Validation/Validator.php]: Good point.

File: src/Validation/Validator.php (hunk 1 above)
Comment (Graziel) [inline src/Validation/Validator.php]: ok thanks!

i suck at grammar : )

**Comment by markstory** [comment]
Looks good to me. @lorenzo ?

File: src/Validation/Validator.php [hunk 4]
```diff
@@ -422,9 +422,16 @@ public function remove($field, $rule = null)
     }
 
     /**
-     * Sets whether a field is required to be present in data array.
+     * Sets whether a field is required to be present in data array. You can also pass array.
```
Comment (ravage84) [inline src/Validation/Validator.php]: I'd move the second sentense ("You can also..."), to the long description.

File: src/Validation/Validator.php [hunk 5]
```diff
@@ -433,10 +440,33 @@ public function remove($field, $rule = null)
...
+            ];
+        }
+
+        foreach ($field as $fieldName => $setting) {
+            if (is_string($setting)) {
+                $fieldName = $setting;
+                $setting = [];
+            }
+            if (!is_array($setting)) {
+                throw new InvalidArgumentException(sprintf('Invalid field "%s" setting, must be an array.', $fieldName));
```
Comment (ravage84) [inline src/Validation/Validator.php]: Nitpicking, you could make that:

``` php
throw new InvalidArgumentException(
    sprintf('Invalid field "%s" setting, must be an array.', $fieldName)
);
````

So it's not that long.

**Comment by lorenzo** [comment]
Thanks @Graziel!

Would you mind adding this to the docs?

**Comment by Graziel** [comment]
@lorenzo sure i can add it to docs but on weekend, too busy right now, 3.next is 3.3 in docs?

@ravage84 sorry i got to much work and didnt make it in time should i make another PR or can you do it?

**Comment by markstory** [comment]
> 3.next is 3.3 in docs?

yes.

**Pull request merged**